### PR TITLE
HAL_ChibiOS: fixed hwdef DMA resolver dependency on python2/python3

### LINKED ISSUE #15534: different hwdef.h for python2 and python3
using python2 vs python3 generates a different hwdef.h for CubeOrange (and likely many other boards). The cause is almost certainly a missing sort in the DMA generator code.

### PR DESCRIPTION
**Pull request by tridge**
we need to use sorted keys in dictionaries to ensure we end up with
the same DMA mapping when using py2 and py3

fixes #15534

ping @WickedShell

### CODE REVIEW

File: Tools/scripts/build_ci.sh [hunk 1]
```diff
@@ -193,7 +193,7 @@ for t in $CI_BUILD_TARGET; do
 
     if [ "$t" == "configure-all" ]; then
         echo "Checking configure of all boards"
-        ./Tools/scripts/configure_all.py
+        ./Tools/scripts/configure_all.py --python=python3 --python2=python2
```
Comment (peterbarker) [inline Tools/scripts/build_ci.sh:196]: Perhaps `--python-a` and `--python-b` would be clearer.

**Review by peterbarker — APPROVED** [review]
Invocation is a bit confusing.

**Comment by davidbuzz** [comment]
So, does this 'sort' bring the current behaviour in-line with the old python2 behaviour,  or the old python3 behaviour,  or were neither of them ok before the change?

**Comment by WickedShell** [comment]
This doesn't bring parity between the old python2 build and the new one with this applied on a CubeOrange. With this applied UART4_Tx is now shared with a bunch of streams, whereas before it was by itself, and I2C2 was shared with other streams. (This is also a bit worrisome as this would quickly reintroduce the baud rate problem I originally hit which triggered the whole investigation here)

``` diff
< #define HAL_VALIDATE_BOARD (!((spi_check_register("icm20948_ext", INV2REG_WHOAMI, INV2_WHOAMI_ICM20948)))?"$CHECK_IMU1_PRESENT":!((spi_check_register("icm20602_ext", MPUREG_WHOAMI, MPU_WHOAMI_ICM20602)))?"$CHECK_IMU0_PRESENT":!((check_ms5611("ms5611_ext")))?"$CHECK_BARO1_PRESENT":!((spi_check_register("icm20948", INV2REG_WHOAMI, INV2_WHOAMI_ICM20649)))?"$CHECK_IMU2_PRESENT":!((check_ms5611("ms5611")))?"$CHECK_BARO0_PRESENT":nullptr)
---
> #define HAL_VALIDATE_BOARD (!((check_ms5611("ms5611")))?"$CHECK_BARO0_PRESENT":!((check_ms5611("ms5611_ext")))?"$CHECK_BARO1_PRESENT":!((spi_check_register("icm20602_ext", MPUREG_WHOAMI, MPU_WHOAMI_ICM20602)))?"$CHECK_IMU0_PRESENT":!((spi_check_register("icm20948_ext", INV2REG_WHOAMI, INV2_WHOAMI_ICM20948)))?"$CHECK_IMU1_PRESENT":!((spi_check_register("icm20948", INV2REG_WHOAMI, INV2_WHOAMI_ICM20649)))?"$CHECK_IMU2_PRESENT":nullptr)
272c272
< #define STM32_I2C_I2C1_TX_DMA_STREAM   STM32_DMA_STREAM_ID(2, 3) // shared UART7_TX,TIM4_UP,I2C2_TX,SPI2_TX,USART3_TX,USART2_TX,I2C1_TX,TIM1_UP
---
> #define STM32_I2C_I2C1_TX_DMA_STREAM   STM32_DMA_STREAM_ID(2, 3) // shared I2C1_TX,SPI2_TX,TIM1_UP,TIM4_UP,UART4_TX,UART7_TX,USART2_TX,USART3_TX
276c276
< #define STM32_I2C_I2C2_TX_DMA_STREAM   STM32_DMA_STREAM_ID(2, 3) // shared UART7_TX,TIM4_UP,I2C2_TX,SPI2_TX,USART3_TX,USART2_TX,I2C1_TX,TIM1_UP
---
> #define STM32_I2C_I2C2_TX_DMA_STREAM   STM32_DMA_STREAM_ID(2, 4)
284c284
< #define STM32_SPI_SPI2_TX_DMA_STREAM   STM32_DMA_STREAM_ID(2, 3) // shared UART7_TX,TIM4_UP,I2C2_TX,SPI2_TX,USART3_TX,USART2_TX,I2C1_TX,TIM1_UP
---
> #define STM32_SPI_SPI2_TX_DMA_STREAM   STM32_DMA_STREAM_ID(2, 3) // shared I2C1_TX,SPI2_TX,TIM1_UP,TIM4_UP,UART4_TX,UART7_TX,USART2_TX,USART3_TX
290c290
< #define STM32_TIM_TIM1_UP_DMA_STREAM   STM32_DMA_STREAM_ID(2, 3) // shared UART7_TX,TIM4_UP,I2C2_TX,SPI2_TX,USART3_TX,USART2_TX,I2C1_TX,TIM1_UP
---
> #define STM32_TIM_TIM1_UP_DMA_STREAM   STM32_DMA_STREAM_ID(2, 3) // shared I2C1_TX,SPI2_TX,TIM1_UP,TIM4_UP,UART4_TX,UART7_TX,USART2_TX,USART3_TX
292c292
< #define STM32_TIM_TIM4_UP_DMA_STREAM   STM32_DMA_STREAM_ID(2, 3) // shared UART7_TX,TIM4_UP,I2C2_TX,SPI2_TX,USART3_TX,USART2_TX,I2C1_TX,TIM1_UP
---
> #define STM32_TIM_TIM4_UP_DMA_STREAM   STM32_DMA_STREAM_ID(2, 3) // shared I2C1_TX,SPI2_TX,TIM1_UP,TIM4_UP,UART4_TX,UART7_TX,USART2_TX,USART3_TX
294c294
< #define STM32_UART_UART4_TX_DMA_STREAM STM32_DMA_STREAM_ID(2, 4)
---
> #define STM32_UART_UART4_TX_DMA_STREAM STM32_DMA_STREAM_ID(2, 3) // shared I2C1_TX,SPI2_TX,TIM1_UP,TIM4_UP,UART4_TX,UART7_TX,USART2_TX,USART3_TX
298c298
< #define STM32_UART_UART7_TX_DMA_STREAM STM32_DMA_STREAM_ID(2, 3) // shared UART7_TX,TIM4_UP,I2C2_TX,SPI2_TX,USART3_TX,USART2_TX,I2C1_TX,TIM1_UP
---
> #define STM32_UART_UART7_TX_DMA_STREAM STM32_DMA_STREAM_ID(2, 3) // shared I2C1_TX,SPI2_TX,TIM1_UP,TIM4_UP,UART4_TX,UART7_TX,USART2_TX,USART3_TX
304c304
< #define STM32_UART_USART2_TX_DMA_STREAM STM32_DMA_STREAM_ID(2, 3) // shared UART7_TX,TIM4_UP,I2C2_TX,SPI2_TX,USART3_TX,USART2_TX,I2C1_TX,TIM1_UP
---
> #define STM32_UART_USART2_TX_DMA_STREAM STM32_DMA_STREAM_ID(2, 3) // shared I2C1_TX,SPI2_TX,TIM1_UP,TIM4_UP,UART4_TX,UART7_TX,USART2_TX,USART3_TX
308c308
< #define STM32_UART_USART3_TX_DMA_STREAM STM32_DMA_STREAM_ID(2, 3) // shared UART7_TX,TIM4_UP,I2C2_TX,SPI2_TX,USART3_TX,USART2_TX,I2C1_TX,TIM1_UP
---
> #define STM32_UART_USART3_TX_DMA_STREAM STM32_DMA_STREAM_ID(2, 3) // shared I2C1_TX,SPI2_TX,TIM1_UP,TIM4_UP,UART4_TX,UART7_TX,USART2_TX,USART3_TX
```

@davidbuzz I discovered when injecting lots of data into the GPS on GPS1 on a CubeOrange, with Telem1 running at a high level as well that my outgoing telemetry was getting massively slowed down. This was eventually tracked down to being a busy DMA stream getting stalled on the slower GPS baud rate, but more horrifyingly the problem goes away if I built with python2 instead of python3, because CubeOrange under python2 had a different allocation of DMA mappings, which makes the builds very inconsistent depending upon your enviorment.

File: Tools/scripts/build_ci.sh (hunk 1 above)
Comment (bugobliterator) [reply Tools/scripts/build_ci.sh:196]: we might want to use a version of python binary that actually isn't named python3 or python2.

**Comment by bugobliterator** [comment]
@tridge I think we need to have a "DMA sharing config" file saved inside hwdef folder as well. This can still be generated, just not generated every time unless the hwdef is changed.

File: Tools/scripts/build_ci.sh (hunk 1 above)
Comment (peterbarker) [reply Tools/scripts/build_ci.sh:196]: @bugobliterator this is simply an invocation of the script which is passing through python3 as primary and python2 as secondary.  That's entirely appropriate for CI, which is what this script is all about :+1:

**Comment by tridge** [comment]
@WickedShell @bugobliterator please see this related PR as well:
https://github.com/ArduPilot/ardupilot/pull/23568

**Review by bugobliterator — APPROVED** [review]

**Pull request merged**